File: java/src/main/java/vsdk/toolkit/raycasting/stochasticRaytracing/Seed.java
package vsdk.toolkit.raycasting.stochasticRaytracing;

/**
SEED Configuration class
*/
public class CSeed {
    private final short[] m_seed;

    public CSeed() {
        m_seed = new short[3];
    }

    public short[] GetSeed() {
        return m_seed;
    }

    public void SetSeed(CSeed seed) {
        short[] s = seed.GetSeed();
        m_seed[0] = s[0];
        m_seed[1] = s[1];
        m_seed[2] = s[2];
    }

    public void SetSeed(short[] seed16v) {
        m_seed[0] = seed16v[0];
        m_seed[1] = seed16v[1];
        m_seed[2] = seed16v[2];
    }

    public void SetSeed(int s0, int s1, int s2) {
        m_seed[0] = (short)s0;
        m_seed[1] = (short)s1;
        m_seed[2] = (short)s2;
    }

    public void XORSeed(CSeed xOrSeed) {
        short[] s = xOrSeed.GetSeed();
        m_seed[0] ^= s[0];
        m_seed[1] ^= s[1];
        m_seed[2] ^= s[2];
    }
}

File: typescript/src/main/vsdk/toolkit/raycasting/stochasticRaytracing/Seed.ts
/**
Configuration class
*/
export class Seed {
  private readonly m_seed: number[];

  public constructor() {
    this.m_seed = [0, 0, 0];
  }

  public GetSeed(): number[] {
    return this.m_seed;
  }

  public SetSeed(seed: Seed): void;
  public SetSeed(seed16v: number[]): void;
  public SetSeed(s0: number, s1: number, s2: number): void;
  public SetSeed(a: Seed | number[] | number, b?: number, c?: number): void {
    if (a instanceof Seed) {
      const s = a.GetSeed();
      this.m_seed[0] = s[0] | 0;
      this.m_seed[1] = s[1] | 0;
      this.m_seed[2] = s[2] | 0;
      return;
    }

    if (Array.isArray(a)) {
      this.m_seed[0] = (a[0] ?? 0) | 0;
      this.m_seed[1] = (a[1] ?? 0) | 0;
      this.m_seed[2] = (a[2] ?? 0) | 0;
      return;
    }

    this.m_seed[0] = a | 0;
    this.m_seed[1] = (b ?? 0) | 0;
    this.m_seed[2] = (c ?? 0) | 0;
  }

  public XORSeed(xOrSeed: Seed): void {
    const s = xOrSeed.GetSeed();
    this.m_seed[0] ^= s[0];
    this.m_seed[1] ^= s[1];
    this.m_seed[2] ^= s[2];
  }
}
